src/rust/ide/lib/view/src/graph_editor/component/breadcrumbs.rs

```diff
@@ -0,0 +1,366 @@
+use std::cmp;
+
+
+
+// =================
+// === LocalCall ===
+// =================
+
+/// A specific function call occurring within another function's definition body.
+/// It's closely related to the `LocalCall` type defined in `Language Server` types, but uses the
+/// new type `MethodPointer` defined in `GraphEditor`.
+#[derive(Clone,Debug,Eq,PartialEq)]
+pub struct LocalCall {
+    /// An expression being a call to a method.
+    pub call:ExpressionId,
+    /// A pointer to the called method.
+    pub definition:MethodPointer,
+}
+
+
+
+// =================
+// === Constants ===
+// =================
+
+// FIXME[dg] hardcoded literal for glyph of height 12.0. Copied from port.rs
+const GLYPH_WIDTH       : f32 = 7.224_609_4;
+const VERTICAL_MARGIN   : f32 = GLYPH_WIDTH;
+const HORIZONTAL_MARGIN : f32 = GLYPH_WIDTH;
+const TEXT_SIZE         : f32 = 12.0;
+
+
+
+// =================
+// === FrpInputs ===
+// =================
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct FrpInputs {
+    /// Push breadcrumb.
+    pub push_breadcrumb:frp::Source<Option<LocalCall>>,
+    /// Pop breadcrumb.
+    pub pop_breadcrumb:frp::Source,
+}
+
+impl FrpInputs {
+    /// Constructor.
+    pub fn new(network:&frp::Network) -> Self {
+        frp::extend! {network
+            push_breadcrumb <- source();
+            pop_breadcrumb  <- source();
+        }
+        Self{push_breadcrumb,pop_breadcrumb}
+    }
+}
+
+
+
+// ==================
+// === FrpOutputs ===
+// ==================
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct FrpOutputs {
+    pub breadcrumb_push : frp::Source<Option<LocalCall>>,
+    pub breadcrumb_pop  : frp::Source
+}
+
+impl FrpOutputs {
+    /// Constructor.
+    pub fn new(network:&frp::Network) -> Self {
+        frp::extend! {network
+            breadcrumb_push <- source();
+            breadcrumb_pop  <- source();
+        }
+        Self{breadcrumb_push,breadcrumb_pop}
+    }
+}
+
+
+
+// ===========
+// === Frp ===
+// ===========
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct Frp {
+    pub inputs  : FrpInputs,
+    pub outputs : FrpOutputs,
+    pub network : frp::Network,
+}
+
+impl Deref for Frp {
+    type Target = FrpInputs;
+    fn deref(&self) -> &Self::Target {
+        &self.inputs
+    }
+}
+
+impl Default for Frp {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Frp {
+    /// Constructor.
+    pub fn new() -> Self {
+        let network = frp::Network::new();
+        let inputs  = FrpInputs::new(&network);
+        let outputs = FrpOutputs::new(&network);
+        Self{network,inputs,outputs}
+    }
+}
+
+
+
+// ========================
+// === BreadcrumbsModel ===
+// ========================
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct BreadcrumbsModel {
+    logger                : Logger,
+    display_object        : display::object::Instance,
+    pub project_name      : ProjectName,
```

Comment:
why this field is public? If you are using FRP you should rather keep such things as FRP outputs. So, for example, if someone wants to react to a new project name, they can just connect to that output. In case you want some 3rd party API allow to questy the project name without connecting to FRP (which I don't like and should not be used but its worth mentioning), you can expose it is `frp::Sampler` - which allows you to call `.value()` and get the value from it without FRP network

Reply:
I made it private and I created new frp endpoints available for integration.